### PR TITLE
Added interaction with local ads dev-servers

### PR DESCRIPTION
**Pull request by rokuz**

### CODE REVIEW

File: map/local_ads_manager.cpp [hunk 1]
```diff
@@ -48,6 +61,50 @@ std::chrono::steady_clock::time_point Now()
 {
   return std::chrono::steady_clock::now();
 }
+
+#ifdef TEMPORARY_LOCAL_ADS_JSON_SERIALIZATION
+struct FreeDeletor
```
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: We already have FreeDeletor in search/search_quality/sample.cpp. Please, move this class to the base and reuse in both places.

File: map/local_ads_manager.cpp [hunk 2]
```diff
@@ -48,6 +61,50 @@ std::chrono::steady_clock::time_point Now()
...
+{
+  ASSERT(!events.empty(), ());
+  contentType = "application/json";
+  contentEncoding = "deflate";
+
+  auto root = my::NewJSONObject();
+  ToJSONObject(*root, "userId", userId);
+  ToJSONObject(*root, "countryId", events.front().m_countryId);
+  ToJSONObject(*root, "mwmVersion", events.front().m_mwmVersion);
+  using namespace std::chrono;
```
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: Please, move this using to the top of the function.

File: map/local_ads_manager.cpp [hunk 3]
```diff
@@ -48,6 +61,50 @@ std::chrono::steady_clock::time_point Now()
...
+    ToJSONObject(*eventNode, "type", static_cast<uint8_t>(event.m_type));
+    ToJSONObject(*eventNode, "timestamp", static_cast<json_int_t>(s));
+    ToJSONObject(*eventNode, "featureId", static_cast<json_int_t>(event.m_featureId));
+    ToJSONObject(*eventNode, "zoomLevel", event.m_zoomLevel);
+    ToJSONObject(*eventNode, "latitude", event.m_latitude);
+    ToJSONObject(*eventNode, "longitude", event.m_longitude);
+    ToJSONObject(*eventNode, "accuracyInMeters", event.m_accuracyInMeters);
+    json_array_append_new(eventsNode.get(), eventNode.release());
+  }
+  json_object_set(root.get(), "events", eventsNode.release());
```
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: If you use release, you should call json_object_set_new, right?

File: map/local_ads_manager.cpp [hunk 4]
```diff
@@ -48,6 +61,50 @@ std::chrono::steady_clock::time_point Now()
 {
   return std::chrono::steady_clock::now();
 }
+
+#ifdef TEMPORARY_LOCAL_ADS_JSON_SERIALIZATION
```
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: Why do we need this ifdef? Why not just to remove this code at appropriate time?

File: local_ads/statistics.cpp [hunk 5]
```diff
@@ -381,21 +384,40 @@ void Statistics::ProcessEvents(std::list<Event> & events)
...
+    std::string contentType = "application/octet-stream";
+    std::string contentEncoding = "";
+    std::vector<uint8_t> bytes = serializer != nullptr
+                                     ? serializer(events, userId, contentType, contentEncoding)
+                                     : SerializeForServer(events);
+    ASSERT(!bytes.empty(), ());
+
     platform::HttpClient request(url);
-    request.SetBodyData(std::string(bytes.begin(), bytes.end()), "application/octet-stream");
+#ifdef DEV_LOCAL_ADS_SERVER
```
Comment (ygorshenin) [inline local_ads/statistics.cpp]: Why do you need these ugly ifdefs? Why not to create a simple C++ class, say, DevLocalAdsServer, that will take the content, prepare headers and run http request?

Comment (rokuz) [inline local_ads/statistics.cpp]: Наличие этого хедера - временное решение. Сергей обещал поднять нормальный dev-сервер, для которого никаких "уродских" хедеров не потребуется

File: map/local_ads_manager.cpp (hunk 4 above)
Comment (rokuz) [inline map/local_ads_manager.cpp]: Это служит указателем тех мест в коде, которые надо удалить, так как места эти идут не подряд.

File: map/local_ads_manager.cpp (hunk 3 above)
Comment (rokuz) [inline map/local_ads_manager.cpp]: +1

File: map/local_ads_manager.cpp (hunk 1 above)
Comment (rokuz) [inline map/local_ads_manager.cpp]: Перенес

File: local_ads/statistics.cpp (hunk 5 above)
Comment (ygorshenin) [inline local_ads/statistics.cpp]: Okay

File: map/local_ads_manager.cpp (hunk 4 above)
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: Okay

**Comment by rokuz** [comment]
PTAL

File: map/local_ads_manager.cpp [hunk 6]
```diff
@@ -48,6 +61,42 @@ std::chrono::steady_clock::time_point Now()
...
+    ToJSONObject(*eventNode, "longitude", event.m_longitude);
+    ToJSONObject(*eventNode, "accuracyInMeters", event.m_accuracyInMeters);
+    json_array_append_new(eventsNode.get(), eventNode.release());
+  }
+  json_object_set_new(root.get(), "events", eventsNode.release());
+  std::unique_ptr<char, JSONFreeDeletor> buffer(
+      json_dumps(root.get(), JSON_COMPACT | JSON_ENSURE_ASCII));
+  std::string buf = buffer.get();
+  std::vector<uint8_t> result;
+  coding::ZLib::Deflate(buf.data(), buf.size(), coding::ZLib::Level::BestCompression,
```
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: Why do we need to copy buffer to buf? You may directly call Deflate:
```c++
Deflate(buffer.get(), strlen(buffer.get()), ...);
```

File: 3party/jansson/myjansson.hpp [hunk 7]
```diff
@@ -106,6 +107,11 @@ void FromJSONObjectOptionalField(json_t * root, std::string const & field, std::
     FromJSON(json_array_get(arr, i), result[i]);
 }
 
+struct JSONFreeDeletor
```
Comment (kshalnev) [inline 3party/jansson/myjansson.hpp]: 'Deleter' is better, see here http://the-difference-between.com/deletor/deleter

File: 3party/jansson/myjansson.hpp [hunk 8]
```diff
@@ -106,6 +107,11 @@ void FromJSONObjectOptionalField(json_t * root, std::string const & field, std::
...
+{
+  void operator()(char * buffer) const { free(buffer); }
```
Comment (kshalnev) [inline 3party/jansson/myjansson.hpp]: May be check buffer is not null? free crashes then.

File: map/local_ads_manager.cpp [hunk 9]
```diff
@@ -48,6 +62,42 @@ std::chrono::steady_clock::time_point Now()
...
+    ToJSONObject(*eventNode, "zoomLevel", event.m_zoomLevel);
+    ToJSONObject(*eventNode, "latitude", event.m_latitude);
+    ToJSONObject(*eventNode, "longitude", event.m_longitude);
+    ToJSONObject(*eventNode, "accuracyInMeters", event.m_accuracyInMeters);
+    json_array_append_new(eventsNode.get(), eventNode.release());
+  }
+  json_object_set_new(root.get(), "events", eventsNode.release());
+  std::unique_ptr<char, JSONFreeDeleter> buffer(
+      json_dumps(root.get(), JSON_COMPACT | JSON_ENSURE_ASCII));
+  std::string buf = buffer.get();
```
Comment (ygorshenin) [inline map/local_ads_manager.cpp]: So, buf is unused, right?

**Comment by rokuz** [comment]
@ygorshenin PTAL

File: 3party/jansson/myjansson.hpp (hunk 7 above)
Comment (rokuz) [inline 3party/jansson/myjansson.hpp]: done

**Pull request merged**